discovery-provider/src/tasks/user_replica_set.py

```diff
@@ -0,0 +1,267 @@
+
+        # If cache miss, fetch from eth-contract
+        if not primary_endpoint:
+            logger.info(f"user_replica_set.py | CACHE MISS FOR {primary_cache_key}, found {primary_info_cached}")
+            if sp_factory_inst is None:
+                sp_factory_inst = get_sp_factory_inst(self, update_task)
+            primary_cn_endpoint_info = sp_factory_inst.functions.getServiceEndpointInfo(
+                content_node_service_type,
+                primary
+            ).call()
+            logger.info(primary_cn_endpoint_info)
+            primary_endpoint = primary_cn_endpoint_info[1]
+
+        endpoint_string = "{}".format(primary_endpoint)
+        for secondary_id in secondaries:
+            secondary_endpoint = None
+            secondary_cache_key = get_sp_id_key(secondary_id)
+            secondary_info_cached = get_pickled_key(update_task.redis, secondary_cache_key)
+            if secondary_info_cached:
+                secondary_endpoint = secondary_info_cached[1]
+                logger.info(f"user_replica_set.py | CACHE HIT FOR {secondary_cache_key}, found {secondary_info_cached}")
+
+            if not secondary_endpoint:
+                logger.info(
+                    f"user_replica_set.py | CACHE MISS FOR {secondary_cache_key}, found {secondary_info_cached}"
+                )
+                if sp_factory_inst is None:
+                    sp_factory_inst = get_sp_factory_inst(self, update_task)
+                secondary_info = sp_factory_inst.functions.getServiceEndpointInfo(
+                    content_node_service_type,
+                    secondary_id
+                ).call()
+                secondary_endpoint = secondary_info[1]
+
+            if secondary_endpoint:
+                endpoint_string = "{},{}".format(endpoint_string, secondary_endpoint)
+            else:
+                logger.info(f"user_replica_set.py | Failed to find secondary info for {secondary_endpoint}")
+    except Exception as exc:
+        logger.error(f"user_replica_set.py | ERROR in get_endpoint_string_from_sp_ids {exc}")
+    logger.info(f"user_replica_set.py | constructed {endpoint_string} from {primary},{secondaries}")
+    return endpoint_string
+
+def get_sp_factory_inst(self, update_task):
+    shared_config = update_task.shared_config
+    eth_web3 = update_task.eth_web3
+    eth_registry_address = eth_web3.toChecksumAddress(
+        shared_config["eth_contracts"]["registry"]
+    )
+    eth_registry_instance = eth_web3.eth.contract(
+        address=eth_registry_address, abi=eth_abi_values["Registry"]["abi"]
+    )
+    sp_factory_address = eth_registry_instance.functions.getContract(
+        sp_factory_registry_key
+    ).call()
+    sp_factory_inst = eth_web3.eth.contract(
+        address=sp_factory_address, abi=eth_abi_values["ServiceProviderFactory"]["abi"]
+    )
+    return sp_factory_inst
+
+def parse_poa_cnode_record(self, update_task, session, entry, cnode_record):
+    event_args = entry["args"]
```

Comment:
nit - I feel like we can direcftly assign cnode_record to have these fields, instead of creating these vars, and then updating cnode_record

Reply:
i think this is better for clarity tbh, good comment but I'm going to stick with the current format